orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -201,45 +277,310 @@ Status TrainingRunner::Run(IDataLoader* training_data_loader, IDataLoader* test_
-  if (loss_scaler_) {
-    feed_names.push_back(loss_scaler_->GetLossScaleInputName());
+  // Pick up feed from loss scaling.
+  if (loss_scaler) {
+    const auto name = loss_scaler->GetLossScaleInputName();
+    if (params_.pipeline_parallel_size == 1 || std::find(allowed_feed_begin, allowed_feed_end, name) != allowed_feed_end) {
+      feed_names.push_back(name);
+      const float loss_scale = loss_scaler->GetLossScale();
+      OrtValue loss_scale_val;
+      TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{loss_scale}, &loss_scale_val, input_allocator_);
+      feeds.push_back(loss_scale_val);
+    }
   }
-  feed_names.push_back(params_.lr_params.feed_name);
 
-  OrtValue loss_scale_val;
-  OrtValue lr_ort_val;
+  // Pick up feed from learning rate schedule.
+  {
+    const auto name = params_.lr_params.feed_name;
+    if (params_.pipeline_parallel_size == 1 || std::find(allowed_feed_begin, allowed_feed_end, name) != allowed_feed_end) {
+      feed_names.push_back(name);
+      // learning rate is 1 if there is no learning-rate scheduler. Otherwise, learning rate is obtained from the scheduler.
+      const float learning_rate = lr_scheduler ? lr_scheduler->GetLearningRate(step_ + 1) : 0.0f;
+      OrtValue lr_val;
+      TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{learning_rate}, &lr_val, input_allocator_);
+      feeds.push_back(lr_val);
+    }
+  }
 
-  VectorString fetch_names = params_.fetch_names;
-  if (params_.use_mixed_precision) {
-    auto it = opt_graph_outputs_.find(OptimizerOutputKey::GradientAllIsFinite);
-    ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Gradient norm's IsFinite output is missing in the optimizer output");
-    fetch_names.push_back(it->second);
-    if (params_.use_adasum) {
-      it = opt_graph_outputs_.find(OptimizerOutputKey::DeltaAllIsFinite);
-      ORT_RETURN_IF(it == opt_graph_outputs_.end(), "Adasum delta's IsFinite output is missing in the optimizer output");
+  // Create feed of waited event in forward pass.
+  if (!pipeline_context_.forward_waited_event_name.empty()) {
+    ORT_ENFORCE(params_.pipeline_parallel_size > 1);
+    feed_names.push_back(pipeline_context_.forward_waited_event_name);
+    OrtValue event_id;
+    const int64_t id = mode == EvaluateStep ? -1 : pipeline_schedule_.GetForwardWaitedEventId(
+      pipeline_context_.pipeline_stage_id,
+      static_cast<int>(step_) % pipeline_context_.num_pipeline_batches);
+    std::cout << id << std::endl;
+    TrainingUtil::CreateCpuMLScalar(
+      id,
+      &event_id,
+      input_allocator_);
+    feeds.push_back(event_id);
+  }
+
+  // Create feed of recorded event in forward pass.
+  if (!pipeline_context_.forward_recorded_event_name.empty()) {
+    ORT_ENFORCE(params_.pipeline_parallel_size > 1);
+    feed_names.push_back(pipeline_context_.forward_recorded_event_name);
+    OrtValue event_id;
+    const int64_t id = mode == EvaluateStep ? -1 : pipeline_schedule_.GetForwardRecordedEventId(
```

Comment:
>GetForwardRecordedEventId [](start = 70, length = 25)

would it better to add a bool to this function for eval mode?

Reply:
I prefer to keep it. Not sure if we will have to do something special for the other two modes in the future.